File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] nge = new int[n];
        Stack<Integer> st = new Stack<>();
        st.push(nums[n - 1]);
        for (int i = n + n - 2; i >= 0; i--) {
            while (!st.isEmpty() && st.peek() <= nums[i % n]) st.pop();
            nge[i % n] = st.isEmpty() ? -1 : st.peek();
            st.push(nums[i % n]);
        }
        return nge;
    }
}

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        vector<int> res(nums.size());
        stack<int> st;
        
        for (int n = nums.size(), i = 2 * n - 1; i >= 0; i--) {
            while (!st.empty() && nums[st.top() % n] <= nums[i % n]) {
                st.pop();
            }
            if (i < n) {
                res[i] = st.empty() ? -1 : nums[st.top() % n];
            }
            st.push(i);
        }
        
        return res;
    }
};